tarodb/assets_store.go

```diff
@@ -1556,6 +1562,44 @@ func (a *AssetStore) LogPendingParcel(ctx context.Context,
 	})
 }
 
+// LogParcelDeliveryAttempt logs a parcel delivery attempt to disk.
+func (a *AssetStore) LogParcelDeliveryAttempt(ctx context.Context,
+	anchorPoint wire.OutPoint) error {
+
+	anchorPointBytes, err := encodeOutpoint(anchorPoint)
+	if err != nil {
+		return err
+	}
+
+	var writeTxOpts AssetStoreTxOptions
+	return a.db.ExecTx(ctx, &writeTxOpts, func(q ActiveAssetsStore) error {
+		// First, we'll fetch the asset transfer using the given
+		// outpoint bytes. We can then relate the delivery attempt to
+		// the asset transfer entry.
+		assetTransfers, err := q.QueryAssetTransfers(ctx, TransferQuery{
+			NewAnchorPoint: anchorPointBytes,
+		})
+		if err != nil {
+			return err
+		}
+		assetTransfer := assetTransfers[0]
```

Comment:
Because `QueryAssetTransfers` is defined as a `:many` in sqlc this query won't return an error when there are no matches for the query (`ErrNoRows`) but an empty slice.

That means that we use an invalid `anchorPoint` we will panic with a `slice bounds out of range`.

We are also doing two queries for storing/retreving these attempts. We go from outpoint->transfer->attempts.
- If outpoint <-> transfer is 1 to 1 we can store outpoint in the `receiver_proof_transfer_attempts` table.
- If one outpoint can be in various transfers (for example after failing to transfer it for first time) then we need an extra field for disambiguating it.

Reply:
I'll raise an error if `len(assetTransfers) == 0`.

> We are also doing two queries for storing/retreving these attempts. We go from outpoint->transfer->attempts.

Do you think it might be reasonable to store the transfer attempt object on the `sendPackage`. The `sendPackage` is carries most of the state which is passed around in the state machine. The asset transfer row is created here:

https://github.com/ffranr/taro/blob/issue164_improve_hashmail_proof_dist/tarodb/assets_store.go#L1481-L1481

Reply:
I don't think that we need the full `sendPackage` obj, but something that indentifies it in one query better than having to first hit the db to go from what we have in the func call  (`anchorPoint`) to `transfer_id` to then query for what we want.

Reply:
> I don't think that we need the full `sendPackage` obj

We already have the the full `sendPackage` obj. What I'm trying to say is that we can add `transfer_id` to that object. And then we don't need to store `anchorPoint` related stuff in the transfer attempts table or query for the `transfer_id`.

Reply:
I think we concluded on our call yesterday that we will leave this for the time being. The optimisation solution requires a complicated SQL statement with several joins. Probably not worth it right now.